### ISSUE TITLE
Suggestion for a new rule: pattern + if -> pattern + guard

### ISSUE THREAD

**Issue by jrp2014** [ndmitchell/hlint#1048]
A suggestion for an additional rule:

```haskell
case ... of
  :
  x -> if p then a else b
```
could, arguably, be better expressed as

```haskell
case ... of
  :
  x | p -> a
  x ->  b
```
I suspect that, in real life, this might not be the endpoint, but could help to unpack and simplify the possible cases and, after simplification, perhaps even bring ghc's pattern completeness checker in to ensure full coverage.

**Comment by shayne-fletcher** [comment]
I think this is pretty but agree that "arguably" applies. I feel it might be a bit strong to globally default enable but it is certainly worth considering as an "opt-in".

**Comment by ndmitchell** [comment]
I think a lot of people carefully craft their pattern matching and if statements to fit their needs, so wouldn't want to change. The example duplicates the `x` pattern. If you switch to `| otherwise` it could even be longer. It's often a good idea, but not in enough cases to make it a global default.

For opt-in hints, I find they are rarely enabled, so I tend not to bother about them as much.

**Comment by jrp2014** [comment]
Thanks.  I can see your point.

Stepping back a bit, just like tuning ghc warnings I suspect that people go with a set that works for a particular project and use them for ci, etc.  But periodically, it can be worth turning on Weverything to see if anything else comes up.  Trouble is that that generates a lot of noise, which is why the warnings were tuned in the first place.

Perhaps there is a case for an audit / diagnostic mode that, just by using a command line flag, no editing of files, switches on a maximalist set of rules and provides only a summary of the rules triggered, with an indication of number of times  / module, etc.  This would allow users to run hlint over their code from time to time, looking for unexpected triggers.  Yes, you can do this manually, but reducing the barrier to entry.  Perhaps that facility already exists, but I didn't spot it.

**Comment by ndmitchell** [comment]
You're looking for the (poorly named) `--show` flag.

**Comment by jrp2014** [comment]
Thanks.  Yes, failed to spot that.  I guess that could be enhanced to produce a summary of the number of times each rule was triggered , to make it easier to spot unexpected hits.

**Comment by ndmitchell** [comment]
If you run `--report` it shows you that information in an HTML UI, which tends to be the easiest way to slice and dice a large number of new errors.

**Comment by jrp2014** [comment]
OK.  Thanks.  I'd used that initially, but forgot about it, for some reason.

**Comment by ndmitchell** [comment]
If you have any good suggestions on how to make this workflow more useful, or easier to find, or whatever you think would help, please let us know.

**Comment by jrp2014** [comment]
A few random thoughts:

*  for getting started, if you are a vim user, vim+ale

**Comment by jrp2014** [comment]
A few random thoughts, most of which could just be captured in the README:

* for getting started, if you are a vim user, vim+ale is easy to set up and a good way of picking up the hints that you have customized
* the .hlint.yaml should be included in cabal's `extra-source-files`, otherwise it will not ship and will not be picked up by `haskell-ci` for travis, for example.
* it may be worth thinking about an example of hlint usage with github actons
* the summary + report approach is useful, but hard to discover and then get into muscle memory.  A more evocative name for the `summary` command would help, as you imply.  `everything` or `unconstrained` or ... might be preferable.  But the approach assumes that you are using a gui, and that you will take the trouble to sift through the report.  Another approach would be to have hlint generate some sort of thumbprint.  Changes to that thumbprint might provide a prompt for further investigation. This need not be complex.  Eg, just generating a list of the (uncontrained) rules triggered + git diff might be enough.  So long as it was easy to work with in the context of continuous integration.

Anyway, documenting how best to use hlnt in the context of ci, to complement the development use case, would be a good first step.

**Comment by josephcsible** [comment]
I'd be in favor of a limited subset of this, where it could fire to reduce duplication. For example, this:
```haskell
case x of
  Just j -> if foo j then bar j else baz
  _ -> baz
```

Would become this:
```haskell
case x of
  Just j | foo j -> bar j
  _ -> baz
```

**Comment by ndmitchell** [comment]
The limited subset seems reasonable. As a first approximation, if you have two alternatives to a case/if tree with the same RHS, you should combine them. Seems tricky to figure out how thought! My guess is if someone can figure it out, its a brand new hint group.

Thanks @jrp2014 for your suggestions:

* I know nothing about Vim+Ale. There's an integrations section in the manual. Want to submit the 1-line summary of what to do or where to find more information?
* HLint with GitHub actions would be great - but I don't know how. Again, I think an HLint/GitHub actions doc/plugin or whatever is required would be great. We have a CircleCI "orb", so one for GitHub actions would be very useful.
* HLint used to be a multi-mode program, so `[lint]` was implicit, but `test` or `grep` were available too. I think that was probably confusing rather than helpful. I've removed `grep` in #1123 and plan to remove `test` and then you'll just be able to run `hlint test` to run it over your test directory.
* I just added a bit to the README about `--show` and `--report`.

I think that deals with all the suggestions in the above post. If someone wants to spin out a separate pattern match adjustment based on deduplication, feel free to open a new ticket.
